contracts/strategies/mstable/IdleMStableStrategy.sol

```diff
@@ -0,0 +1,237 @@
+
+        vault.stake(interestTokenAvailable);
+
+        _mint(msg.sender, interestTokenAvailable);
+        return interestTokenAvailable;
+    }
+
+    // _amount is strategy token
+    function redeem(uint256 _amount) external override onlyIdleCDO returns (uint256) {
+        return _redeem(_amount);
+    }
+
+    // _amount in underlying token
+    function redeemUnderlying(uint256 _amount) external override returns (uint256) {
+        uint256 _underlyingAmount = (_amount * oneToken) / price();
+        return _redeem(_underlyingAmount);
+    }
+
+    function getApr() external view override returns (uint256) {
+        uint256 rawBalance = vault.rawBalanceOf(address(this));
+        uint256 expectedUnderlyingAmount = imUSD.creditsToUnderlying(rawBalance);
+
+        uint256 gain = expectedUnderlyingAmount - lastIndexAmount;
+        if (gain == 0) {
+            return 0;
+        }
+        uint256 time = block.timestamp - lastIndexedTime;
+        uint256 gainPerc = (gain * 10**20) / lastIndexAmount;
+        uint256 apr = (YEAR / time) * gainPerc;
+        return apr;
+    }
+
+    /* -------- internal functions ------------- */
+
+    // here _amount means credits, will redeem any governance token if there
+    function _redeem(uint256 _amount) internal returns (uint256) {
+        require(_amount != 0, "Amount shuld be greater than 0");
+        
+        lastIndexAmount = lastIndexAmount - _amount;
+        lastIndexedTime = block.timestamp;
+
+        _burn(msg.sender, _amount);
+        vault.withdraw(_amount);
+
+        uint256 massetReceived = imUSD.redeem(_amount);
+        underlyingToken.transfer(msg.sender, massetReceived);
+
+        return massetReceived;
+    }
+
+    function _swapGovTokenOnUniswapAndDepositToVault(uint256 minLiquidityTokenToReceive) internal returns (uint256) {
+        uint256 govTokensToSend = IERC20Detailed(govToken).balanceOf(address(this));
+
+        IERC20Detailed(govToken).approve(address(uniswapV2Router02), govTokensToSend);
+
+        uint256 underlyingTokenBalanceBefore = underlyingToken.balanceOf(address(this));
+        uniswapV2Router02.swapExactTokensForTokens(
+            govTokensToSend,
+            minLiquidityTokenToReceive,
+            uniswapRouterPath,
+            address(this),
+            block.timestamp
+        );
+        uint256 underlyingTokenBalanceAfter = underlyingToken.balanceOf(address(this));
+
+        require(
+            underlyingTokenBalanceAfter - underlyingTokenBalanceBefore >= minLiquidityTokenToReceive,
+            "Should received more reward from uniswap than minLiquidityTokenToReceive"
+        );
+
+        uint256 underlyingBalanceAfter = underlyingToken.balanceOf(address(this));
+        uint256 newCredits = _depositToVault(underlyingBalanceAfter);
+        _mint(msg.sender, newCredits);
+        return newCredits;
+    }
+
+    function claimGovernanceTokens(uint256 startRound, uint256 endRound) public onlyOwner {
```

Comment:
Why is this necessary? Harvests should claim governance tokens when called, this is adding an additional step to this strategy harvesting.

Reply:
I think `vault.claimRewards()` may take up infinite gas sometimes and may lead to contracts not able to withdraw META tokens. Hence to prevent it I added a function `claimGovernanceTokens(uint256 startRound, uint256 endRound)` which calls `vault.claimRewards(startRound, endRound)`;

Let me know if my thought is wrong, and I will correct the contract accordingly.

Reply:
what about saving and updating a variable `lastClaimedRound` upon initialize and then on each `claimGovernanceTokens` ?

eg something like this 
```
uint256 endRound = fetchEndRound();
_claimGovernanceTokens(lastClaimedRound, endRound);
lastClaimedRound = endRound;
```
assuming there is a way to fetch the current round number via `fetchEndRound`

Reply:
The rounds are expected to be calculated offchain and sent as params to `function redeemRewards(bytes calldata _extraData)` 
The function is only callable be idleCDO, hence it this approach.

Reply:
Yes but how? If we can return those value directly in the redeemRewards we can do a static call without any other additional work off-chain basically. If we can keep track of the last claimed round on-chain we probably don't need to pass anything calculated off-chain which will be handy, wdyt?

Reply:
If pass input `(0,0)` to input arguments, the contract will claim all the possible tokens available. The vault contract has functions to solve this.

In short `claimGovernance(0,0)` with withdraw all possible tokens
`claimGoveranance(a,b)` will claim from round a to round b.

https://github.com/idlehusbandry/idle-tranches/blob/2fba9dddab303713cfb0313db268d558eed6e135/contracts/strategies/mstable/IdleMStableStrategy.sol#L268